src/stackdriver-nozzle/firehose/client.go

```diff
@@ -25,30 +30,41 @@ func NewClient(apiAddress, username, password string, skipSSLValidation bool) Cl
 			SkipSslValidation: skipSSLValidation}}
 }
 
-func (fc *client) StartListening(nozzle logging.Logging) error {
-	cfClient := cfclient.NewClient(fc.cfConfig)
+func (c *client) StartListening(fh FirehoseHandler) error {
+	cfConfig := &cfclient.Config{
+		ApiAddress:        c.cfConfig.ApiAddress,
+		Username:          c.cfConfig.Username,
+		Password:          c.cfConfig.Password,
+		SkipSslValidation: c.cfConfig.SkipSslValidation}
+	cfClient := cfclient.NewClient(cfConfig)
 
-	cachingClient := caching.NewCachingEmpty()
+	cfConsumer := consumer.New(
+		cfClient.Endpoint.DopplerEndpoint,
+		&tls.Config{InsecureSkipVerify: c.cfConfig.SkipSslValidation},
+		nil)
 
-	eventRouter := eventRouting.NewEventRouting(cachingClient, nozzle)
-
-	//err := eventRouter.SetupEventRouting("LogMessage,ValueMetric,HttpStartStop,CounterEvent,Error,ContainerMetric")
-	err := eventRouter.SetupEventRouting("HttpStartStop,ValueMetric")
-	if err != nil {
-		return err
-	}
+	refresher := CfClientTokenRefresh{cfClient: cfClient}
+	cfConsumer.SetIdleTimeout(time.Duration(30) * time.Second)
+	cfConsumer.RefreshTokenFrom(&refresher)
+	messages, errs := cfConsumer.Firehose("test", "")
 
-	if nozzle.Connect() {
-		firehoseConfig := &firehoseclient.FirehoseConfig{
-			TrafficControllerURL:   cfClient.Endpoint.DopplerEndpoint,
-			InsecureSSLSkipVerify:  fc.cfConfig.SkipSslValidation,
-			IdleTimeoutSeconds:     30,
-			FirehoseSubscriptionID: "stackdriver-nozzle",
+	for {
+		select {
+		case envelope := <-messages:
+			err := fh.HandleEvent(envelope)
+			if err != nil {
+				return err
+			}
+		case err := <-errs:
```

Comment:
This ensures we don't jump out as soon as we see the first error while firehose is still trying to send it through. 

The only oddity is if an error occurs at some point in the stream but is recovered from, we will return the last error from this function on normal exit. This might be a case where we want to just log the error channel out to stderr of the nozzle and continue to drain the messages channel till it's closed.